### PR TITLE
Bugfixes and improvements

### PR DESCRIPTION
**Pull request by dominikl**
Some bugfixes and improvements for the Java Gateway.

Previously was https://github.com/ome/omero-blitz/pull/16 and also even earlier https://github.com/openmicroscopy/openmicroscopy/pull/5935

FYI: PR for integration test is https://github.com/openmicroscopy/openmicroscopy/pull/6022

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -13,7 +13,7 @@ repositories {
 }
 
 dependencies {
-    api("org.openmicroscopy:omero-blitz:5.5.0-m4")
+    api("org.openmicroscopy:omero-blitz:5.5.0-SNAPSHOT")
```
Comment (jburel) [inline build.gradle]: should be moved to m5 not SNAPSHOT

File: build.gradle [hunk 2]
```diff
@@ -13,7 +13,7 @@ repositories {
 }
 
 dependencies {
-    api("org.openmicroscopy:omero-blitz:5.5.0-SNAPSHOT")
+    api("org.openmicroscopy:omero-blitz:5.5.0-m4")
```
Comment (jburel) [inline build.gradle]: m5

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [OMERO-build-push#106](https://web-proxy.openmicroscopy.org/west-ci/job/OMERO-build-push/106/). See the [console output](https://web-proxy.openmicroscopy.org/west-ci/job/OMERO-build-push/106/consoleText) for more details.
Possible conflicts:
  - PR #1 dominikl 'gradle settings'
    - build.gradle
  - Upstream changes
    - build.gradle

--conflicts

**Comment by jburel** [comment]
https://github.com/ome/omero-java-gateway/pull/3/commits/e2ffb760de6edc6cf41e22d5080ab528bce2c99d should be extracted from this PR 
so we can tag m3 without having to review the full set of fixes

File: src/main/java/omero/gateway/model/ROIData.java [hunk 3]
```diff
@@ -244,29 +244,48 @@ public int getShapeCount()
...
+        }
+        if (t != -1) {
+            List<ShapeData> allT = roiShapes.get(new ROICoordinate(z, -1));
+            if (allT != null)
+                res.addAll(allT);
+        }
         return res;
     }
 
+
```
Comment (mtbc) [inline src/main/java/omero/gateway/model/ROIData.java]: Unintended?

File: src/main/java/omero/gateway/util/Links.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    /**
+     * Set the parent, child or both of an existing link
+     *
+     * @param link   The link object
+     * @param parent The parent (or <code>null</code>)
+     * @param child  The child (or <code>null</code>)
+     * @return The link object
+     * @throws ClassNotFoundException
```
Comment (mtbc) [inline src/main/java/omero/gateway/util/Links.java]: Possibly not worth these throws lines if there's no actual documentation! Also may want to consider just brushing under the `ReflectiveOperationException` superclass.

File: src/main/java/omero/gateway/util/Mask.java [hunk 5]
```diff
@@ -157,15 +164,72 @@ private static int getFirstNonZeroInt(int[][] array) {
...
+     * int[] x = 0 1 2 3 4 5 6 7 8 9
+     *
+     * using length = 4 will be transformed to
+     *
+     * int[][] y =
+     * [0 1 2 3]
+     * [4 5 6 7]
+     * [8 9 0 0]
+     * </pre>
+     * @param array  An int array
```
Comment (mtbc) [inline src/main/java/omero/gateway/util/Mask.java]: double space

**Comment by mtbc** [comment]
Would it be possible to add some tests for the `ROIData.getShapes` codepaths?

File: src/main/java/omero/gateway/model/ROIData.java [hunk 6]
```diff
@@ -244,29 +244,48 @@ public int getShapeCount()
...
+        List<ShapeData> res = new ArrayList<ShapeData>();
+        roiShapes.values().stream().forEach(list -> res.addAll(list));
+        return res;
+    }
+
+    /**
+     * Returns a list of shapes on a given plane.
+     *
+     * @param z
+     *            The z-section.
```
Comment (mtbc) [inline src/main/java/omero/gateway/model/ROIData.java]: Could add docs on how "-1" is special.

**Comment by dominikl** [comment]
> Would it be possible to add some tests for the `ROIData.getShapes` codepaths?
That's true. I'll add a unit test.

File: src/main/java/omero/gateway/util/Links.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
+package omero.gateway.util;
```
Comment (jburel) [inline src/main/java/omero/gateway/util/Links.java]: needs copyright header

File: src/main/java/omero/gateway/util/Mask.java [hunk 8]
```diff
@@ -157,15 +164,72 @@ private static int getFirstNonZeroInt(int[][] array) {
...
+        int[][] result = new int[height][length];
+        for (int i = 0; i < array.length; i++)
+            result[i / length][i % length] = array[i];
+        return result;
+    }
+
+    /**
+     * Transforms a 2 dimensional array into a one dimensional
+     * array; the opposite of the fold method.
+     * E. g.
```
Comment (jburel) [inline src/main/java/omero/gateway/util/Mask.java]: E.g. extra space

**Comment by dominikl** [comment]
Added some unit tests for the `ROIData.getShapes` methods @mtbc . Are they actually run by travis (/gradle)? Can't see anything in the output.

**Comment by jburel** [comment]
@dominikl unit test should be run by travis see `` ./gradlew test`` in the travis file

**Comment by dominikl** [comment]
Ok, so maybe there's just no detailed output about the tests. However i.e. if travis is green == tests passed as well.

**Comment by jburel** [comment]
yes I created an issue  a month ago on one repo only to print out the output
https://github.com/ome/omero-common/issues/16

File: src/main/java/omero/gateway/facility/DataManagerFacility.java [hunk 9]
```diff
@@ -807,5 +809,56 @@ public FileAnnotationData call() throws Exception {
         }
         return null;
     }
+
+    /**
+     * Moves DataObjects from one object to another, i.e. Annotations from one
+     * Image to another, or Images from one Dataset to another.
```
[the remark this answers is not in the document]
Comment (jburel) [reply src/main/java/omero/gateway/facility/DataManagerFacility.java]: That can be done in a follow up PR

**Comment by joshmoore** [comment]
Thanks, @jburel . Rolling into 5.5.0-m6 then.

**Pull request merged**